File: APPLICATION/services/cliente_service.py
"""
cliente_service.py — Servicio CRUD para clientes.
Toda persistencia se delega al FileManager (struct + seek()).
"""
import csv
from typing import List, Optional

from models.cliente import Cliente
from services.file_manager import FileManager
from utils.config import CLIENTES_DAT, CLIENTES_IDX, CLIENTES_HOLES
from utils.logger import get_logger

logger = get_logger("cliente_service")


class ClienteService:
    """Operaciones CRUD sobre clientes usando archivos binarios de acceso directo."""

    def __init__(self) -> None:
        self._fm = FileManager(
            dat_path=CLIENTES_DAT,
            idx_path=CLIENTES_IDX,
            holes_path=CLIENTES_HOLES,
            model_class=Cliente,
        )

    # ── Crear ────────────────────────────────────────────────────────────────

    def crear(self, nombre: str, dni: str, telefono: str = "", email: str = "") -> Cliente:
        """Registra un nuevo cliente."""
        if not nombre.strip():
            raise ValueError("El nombre del cliente no puede estar vacío.")
        if not dni.strip():
            raise ValueError("El DNI no puede estar vacío.")
        # Verificar DNI duplicado
        if self.buscar_por_dni(dni):
            raise ValueError(f"Ya existe un cliente con DNI {dni}.")

        cliente = Cliente(
            nombre=nombre.strip(), dni=dni.strip(),
            telefono=telefono.strip(), email=email.strip(), estado=1,
        )
        self._fm.insert(cliente)
        logger.info(f"Cliente creado: {cliente}")
        return cliente

    # ── Leer ─────────────────────────────────────────────────────────────────

    def obtener(self, cliente_id: int) -> Optional[Cliente]:
        """Acceso directo O(1) por ID."""
        return self._fm.find_by_id(cliente_id)

    def listar(self) -> List[Cliente]:
        """Todos los clientes activos."""
        return self._fm.find_all()

    def buscar(self, termino: str) -> List[Cliente]:
        """Búsqueda por nombre o DNI."""
        t = termino.lower().strip()
        return [
            c for c in self._fm.find_all()
            if t in c.nombre.lower() or t in c.dni.lower()
        ]

    def buscar_por_dni(self, dni: str) -> Optional[Cliente]:
        """Busca un cliente específico por su DNI."""
        dni_clean = dni.strip()
        for c in self._fm.find_all():
            if c.dni == dni_clean:
                return c
        return None

    # ── Actualizar ───────────────────────────────────────────────────────────

    def actualizar(
        self,
        cliente_id: int,
        nombre:   Optional[str] = None,
        dni:      Optional[str] = None,
        telefono: Optional[str] = None,
        email:    Optional[str] = None,
    ) -> Optional[Cliente]:
        """Modifica campos de un cliente existente."""
        cli = self._fm.find_by_id(cliente_id)
        if cli is None:
            logger.warning(f"Cliente id={cliente_id} no encontrado para actualizar")
            return None

        if nombre   is not None: cli.nombre   = nombre.strip()
        if dni      is not None: cli.dni      = dni.strip()
        if telefono is not None: cli.telefono = telefono.strip()
        if email    is not None: cli.email    = email.strip()

        self._fm.update(cli)
        logger.info(f"Cliente actualizado: {cli}")
        return cli

    # ── Eliminar ─────────────────────────────────────────────────────────────

    def eliminar(self, cliente_id: int) -> bool:
        """Eliminación lógica (estado=0)."""
        ok = self._fm.delete(cliente_id)
        if ok:
            logger.info(f"Cliente id={cliente_id} eliminado lógicamente")
        return ok

    # ── Utilidades ───────────────────────────────────────────────────────────

    def contar(self) -> int:
        return self._fm.count_active()

    def exportar_csv(self, export_path) -> int:
        """Exporta todos los clientes activos a CSV para sincronización."""
        clientes = self.listar()
        with open(export_path, "w", encoding="utf-8", newline="") as f:
            writer = csv.writer(f)
            writer.writerow(["id", "nombre", "dni", "telefono", "email", "estado"])
            for c in clientes:
                writer.writerow([c.id, c.nombre, c.dni,
                                  c.telefono, c.email, c.estado])
        logger.info(f"Exportados {len(clientes)} clientes a {export_path}")
        return len(clientes)

File: APPLICATION-JAVA/src/main/java/com/minimarket/services/ClienteService.java
package com.minimarket.services;

import com.minimarket.models.Cliente;
import com.minimarket.utils.Config;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.List;
import java.util.logging.Logger;
import java.util.stream.Collectors;

/**
 * Servicio CRUD para clientes.
 * Toda la persistencia se delega al FileManager (RandomAccessFile + seek()).
 */
public class ClienteService {

    private static final Logger LOG = Logger.getLogger(ClienteService.class.getName());

    private final FileManager<Cliente> fm;

    public ClienteService() {
        fm = new FileManager<>(
            Config.CLIENTES_DAT,
            Config.CLIENTES_IDX,
            Config.CLIENTES_HOLES,
            Cliente.SIZE,
            Cliente::fromBytes,
            "Cliente"
        );
    }

    // ── Crear ────────────────────────────────────────────────────────────────

    public Cliente crear(String nombre, String dni, String telefono, String email) {
        if (nombre == null || nombre.isBlank())
            throw new IllegalArgumentException("El nombre del cliente no puede estar vacío.");
        if (dni == null || dni.isBlank())
            throw new IllegalArgumentException("El DNI no puede estar vacío.");
        if (buscarPorDni(dni) != null)
            throw new IllegalArgumentException("Ya existe un cliente con DNI " + dni + ".");

        Cliente cliente = new Cliente(0, nombre.strip(), dni.strip(),
            telefono != null ? telefono.strip() : "",
            email    != null ? email.strip()    : "", 1);
        fm.insert(cliente);
        LOG.info("Cliente creado: " + cliente);
        return cliente;
    }

    // ── Leer ─────────────────────────────────────────────────────────────────

    /** Acceso directo O(1) por ID. */
    public Cliente obtener(int clienteId) {
        return fm.findById(clienteId);
    }

    /** Todos los clientes activos. */
    public List<Cliente> listar() {
        return fm.findAll();
    }

    /** Busqueda por nombre o DNI. */
    public List<Cliente> buscar(String termino) {
        String t = termino.toLowerCase().strip();
        return fm.findAll().stream()
            .filter(c -> c.getNombre().toLowerCase().contains(t)
                      || c.getDni().toLowerCase().contains(t))
            .collect(Collectors.toList());
    }

    /** Busca un cliente especifico por su DNI. */
    public Cliente buscarPorDni(String dni) {
        String dniClean = dni.strip();
        return fm.findAll().stream()
            .filter(c -> c.getDni().equals(dniClean))
            .findFirst()
            .orElse(null);
    }

    // ── Actualizar ───────────────────────────────────────────────────────────

    public Cliente actualizar(int clienteId, String nombre, String dni,
                              String telefono, String email) {
        Cliente cli = fm.findById(clienteId);
        if (cli == null) {
            LOG.warning("Cliente id=" + clienteId + " no encontrado para actualizar");
            return null;
        }
        if (nombre   != null) cli.setNombre(nombre.strip());
        if (dni      != null) cli.setDni(dni.strip());
        if (telefono != null) cli.setTelefono(telefono.strip());
        if (email    != null) cli.setEmail(email.strip());

        fm.update(cli);
        LOG.info("Cliente actualizado: " + cli);
        return cli;
    }

    // ── Eliminar ─────────────────────────────────────────────────────────────

    /** Eliminacion logica (estado=0). */
    public boolean eliminar(int clienteId) {
        boolean ok = fm.delete(clienteId);
        if (ok) LOG.info("Cliente id=" + clienteId + " eliminado logicamente");
        return ok;
    }

    // ── Utilidades ───────────────────────────────────────────────────────────

    public int contar() { return fm.countActive(); }

    /** Exporta todos los clientes activos a CSV para sincronizacion. */
    public int exportarCsv(Path exportPath) throws IOException {
        List<Cliente> clientes = listar();
        try (PrintWriter pw = new PrintWriter(Files.newBufferedWriter(exportPath, StandardCharsets.UTF_8))) {
            pw.println("id,nombre,dni,telefono,email,estado");
            for (Cliente c : clientes) {
                pw.println(c.toCsvRow());
            }
        }
        LOG.info("Exportados " + clientes.size() + " clientes a " + exportPath);
        return clientes.size();
    }
}
